packages/wouter/src/memory-location.js

```diff
@@ -0,0 +1,50 @@
+import mitt from "mitt";
+import { useSyncExternalStore } from "./react-deps.js";
+import { relativePath } from "./paths.js";
+
+/**
+ * In-memory location that supports navigation
+ */
+
+export const memoryLocation = ({
+  path = "/",
+  static: staticLocation,
+  record,
+} = {}) => {
+  let currentPath = path;
+  const history = [currentPath];
+  const emitter = mitt();
+
+  const navigateImplementation = (path, { replace = false } = {}) => {
+    if (record) {
+      if (replace) {
+        history.splice(history.length - 1, 1, path);
+      } else {
+        history.push(path);
+      }
+    }
+
+    currentPath = path;
+    emitter.emit("navigate", path);
+  };
+
+  const navigate = !staticLocation ? navigateImplementation : () => null;
+
+  const useMemoryLocation = ({ base } = {}) => {
+    const location = useSyncExternalStore(
+      (cb) => {
+        emitter.on("navigate", cb);
+        return () => emitter.off("navigate", cb);
+      },
```

Comment:
Oh, that's a common caveat when working with `uSES`, I had this problem once too. React will resubscribe every time the first argument changes, so the reference to this function should stay the same to avoid extra work.

> The store snapshot returned by getSnapshot must be immutable. If the underlying store has mutable data, return a new immutable snapshot if the data has changed. Otherwise, return a cached last snapshot.